### ISSUE TITLE
declare_interior_mutable_const shouldn't fire for const thread-local initializers

### ISSUE THREAD

**Issue by goffrie** [rust-lang/rust-clippy#8493]
### Summary

`thread_local!` has an optimization for `const` initializers. Also, thread-locals often have interior mutability. When using these two features together, the `declare_interior_mutable_const` lint fires, as it detects an interior-mutable `const` item. Which is true, but it is not being used in an incorrect manner; the use happens inside of the `thread_local!` macro.

### Lint Name

declare_interior_mutable_const

### Reproducer

```rust
use std::cell::Cell;
thread_local! {
    static THREAD_LOCAL: Cell<i32> = const { Cell::new(0) };
}
```

I saw this happen:

```
warning: a `const` item should never be interior mutable
 --> src/lib.rs:2:1
  |
2 | / thread_local! {
3 | |     static THREAD_LOCAL: Cell<i32> = const { Cell::new(0) };
4 | | }
  | |_^
  |
  = note: `#[warn(clippy::declare_interior_mutable_const)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#declare_interior_mutable_const
  = note: this warning originates in the macro `$crate::__thread_local_inner` (in Nightly builds, run with -Z macro-backtrace for more info)
```

### Version

```text
rustc 1.60.0-nightly (bd3cb5256 2022-01-16)
binary: rustc
commit-hash: bd3cb52565faab2755ff1bdb54d88bc91f47b4b9
commit-date: 2022-01-16
host: x86_64-apple-darwin
release: 1.60.0-nightly
LLVM version: 13.0.0
```

**Comment by hellow554** [comment]
The problem seems, that #9015 only covers an unexpanded `thread_local`, so here's a MCVE that still triggers the lint:

```rust
macro_rules! task_local{
    () => {
        static _FOO: () = {
            std::thread_local! {
                static __KEY: std::cell::RefCell<Option<()>> = const { std::cell::RefCell::new(None) };
            }
        };
    };
}

task_local!();

fn main() {}
```

@kyoto7250 may I ask you to have another look at this?

**Comment by kyoto7250** [comment]
@hellow554 
Thank you for your comment.

I was able to reproduce it.


---

@Jarcho 

I would like to hear your thought.

Should we expand `MacroCall` and check everything out?

**Comment by Jarcho** [comment]
You'll have to check every item in the macro backtrace since it will be somewhere in the middle.
